### PR TITLE
OEP-48: Brand Customization

### PR DESCRIPTION
**Pull request by abutterworth**
We will create a new project that defines a branding package interface and a default implementation for branding and theming Open edX applications. The brand package will have the capability to customize assets referred to in the SASS theme like main colors, fonts, and particular backgrounds. It will also include specific image assets such as logo and favicon.

### CODE REVIEW

**Review by xitij2000 — COMMENTED** [review]
Should this perhaps also mention `frontend-component-header` etc that can also be replaced with aliases?

I think I've seen it mentioned elsewhere but this might be a good place as well.

File: oeps/oep-0048-brand-customization.rst [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+
+Context
+-------
+
+The ability to customize the Open edX platform to reflect a custom brand and visual style is a critically important feature for the Open edX community. The comprehensive theming system built into edx-platform has served theme and branding needs for several years. It is imperfect and unofficially supported, but it has worked.
+
+Meanwhile, the edX organization has invested in a new microfrontend architecture that features many small, independently deployable microfrontend applications (MFEs). Most of the Open edX user interface will be delivered by these MFEs by the end of this technical transition.
+
+The UI in MFEs are primarily built using the Paragon SCSS and React library based on Bootstrap 4. Paragon itself supports themes via SCSS variable definitions in the same way as its parent library, Bootstrap. The edx.org theme lives inside the Paragon project along side a starter theme.
```
[the remark this answers is not in the document]
Comment (xitij2000) [reply oeps/oep-0048-brand-customization.rst:38]: Additional note: Using a JSON-only theme is also harder to support on edx-platform.

**Comment by xitij2000** [comment]
Note: I will be on leave till 15th October, and will look at the developments at the earliest after I'm back.

File: oeps/oep-0048-brand-customization.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (abutterworth) [reply oeps/oep-0048-brand-customization.rst:38]: I think we should continue our consideration of using a CSS-in-JS theming system. If we do adopt one, it should live in a brand package aside the SCSS theming system (since we're unlikely to replace it in a short period).

File: oeps/oep-0048-brand-customization.rst [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+Today, MFEs hardcode the inclusion of the edx.org Paragon theme. Custom themes and branding are difficult to apply. Why is this the case?
+
+A key distinction between the MFEs edX has created and server-rendered pages is that MFEs are built, static applications. It may be better to think of MFEs as "web app generators" rather than web applications themselves. They are given a configuration as input and they output static html, javascript, and css that is then deployed to the web.
+
+Whereas a server application can read a configuration file when a user requests a page, an edX MFE does not. A pipeline builds and deploys the MFE with a configuration. If the configuration changes, the pipeline must be rerun and the MFE redeployed.
```
[the remark this answers is not in the document]
Comment (abutterworth) [reply oeps/oep-0048-brand-customization.rst:44]: Agreed. A couple other runtime theming possibilities:

- If we transition our CSS to use CSS variables (not supported in IE 11) we'll be able to swap themes at runtime as well. 
-  We could load style sheets using a frontend plugin system, and choose from several precompiled sheets + themes.

Though I think this description accurately to describes how our system currently functions. Maybe I should change: "the MFEs edX has created" to "Open edX MFEs"

File: oeps/oep-0048-brand-customization.rst [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+We will create a new project that defines a branding package interface and a default implementation for branding and theming Open edX applications. The brand package will have the capability to customize assets referred to in the SASS theme like main colors, fonts, and particular backgrounds. It will also include specific image assets such as logo and favicon.
+
+A branding package will be a node module that contains a defined set of files and directories. The default implementation will be available on Github at `edx/brand-openedx <https://github.com/edx/openedx>`_ and published to npm as `@edx/brand-openedx`.
+
+Brand packages will be installed in edx-platform and all MFEs via npm, aliased as `@edx/brand`.
+
+.. image:: oep-0048/branding_architecture_overview.png
+
+Those who wish to customize the branding of their Open edX installation, including edx.org, will create a package that implements this branding package interface. The edx.org branding package will be made public on github at `edx/brand-edx.org <https://github.com/edx/brand-edx.org>`_.
```
[the remark this answers is not in the document]
Comment (abutterworth) [reply oeps/oep-0048-brand-customization.rst:63]: I agree with this. Paragon should supply defaults for any properties it requires. It should have no dependency on a theme.

File: oeps/oep-0048-brand-customization.rst [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+As MFEs serve more and more of the platform user interface, a system for branding and theming them must be created. An ideal solution would apply to both MFEs and edx-platform.
+
+Decision
+--------
+
+We will create a new project that defines a branding package interface and a default implementation for branding and theming Open edX applications. The brand package will have the capability to customize assets referred to in the SASS theme like main colors, fonts, and particular backgrounds. It will also include specific image assets such as logo and favicon.
+
+A branding package will be a node module that contains a defined set of files and directories. The default implementation will be available on Github at `edx/brand-openedx <https://github.com/edx/openedx>`_ and published to npm as `@edx/brand-openedx`.
+
+Brand packages will be installed in edx-platform and all MFEs via npm, aliased as `@edx/brand`.
```
[the remark this answers is not in the document]
Comment (abutterworth) [reply oeps/oep-0048-brand-customization.rst:59]: In the near term this brand package system will need to live together with comprehensive theming. I don't have a clear picture of how brand package assets will be included in a comprehensive theme, but I don't think we should change much here. I think comprehensive themes should leverage brand package assets where possible, but I don't think we should force any dependencies here.

**Comment by OmarIthawi** [comment]
Thanks @abutterworth for driving this. I'd like to understand better how this would impact multi-tenant deployments. 

 - **Theming at build time:** Is the proposal mandating a dedicated MFE installation per Site? or
 - **Theming at runtime:** the package could be dynamic to pull that information form an API.

If it's the latter it would be much better, otherwise I'd like to know how would you plan to manage it?

**Comment by abutterworth** [comment]
> Thanks @abutterworth for driving this. I'd like to understand better how this would impact multi-tenant deployments.
> 
> * **Theming at build time:** Is the proposal mandating a dedicated MFE installation per Site? or
> * **Theming at runtime:** the package could be dynamic to pull that information form an API.
> 
> If it's the latter it would be much better, otherwise I'd like to know how would you plan to manage it?

@OmarIthawi  In our MFEs today, themes are applied with SCSS when the MFE builds. If you have multiple themes you'll need to build the MFE once for every theme you have. This proposal is not a recommendation to mandating this. However, the usage it describes currently is "theming at build time". A brand package should be agnostic to how the assets it contains are consumed, whether at build time or runtime. We don't currently have the technology built to do runtime theming of MFEs. For changes on this front it may be useful to track work being done on frontend plugins (@davidjoy) and MFE build pipelines (https://github.com/felipemontoya).

File: oeps/oep-0048-brand-customization.rst [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+Today, MFEs hardcode the inclusion of the edx.org Paragon theme. Custom themes and branding are difficult to apply. Why is this the case?
+
+A key distinction between the MFEs created for Open edX and server-rendered pages is that MFEs are built, static applications. It may be better to think of MFEs as "web app generators" rather than web applications themselves. They are given a configuration as input and they output static html, javascript, and css that is then deployed to the web.
+
+Whereas a server application can read a configuration file when a user requests a page, an Open edX MFE does not. A pipeline builds and deploys the MFE with a configuration. If the configuration changes, the pipeline must be rerun and the MFE redeployed.
+
+MFEs lack the kind configurability available in Open edX's server applications because edX does not share a common build and deploy system with the community. Work to create an MFE build pipeline for the community is currently underway.
```
Comment (davidjoy) [inline oeps/oep-0048-brand-customization.rst:46]: Maybe worth noting that this work is outside the scope of this OEP. ¯\_(ツ)_/¯

File: oeps/oep-0048-brand-customization.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidjoy) [reply oeps/oep-0048-brand-customization.rst:63]: I also think this makes sense. 👍   I was envisioning the brand-openedx repo as more of a template that explains _how_ to create your brand, maybe with empty, sample files in place with comments in them explaining what to do.

**Comment by davidjoy** [comment]
@OmarIthawi I have similar thoughts and concerns, and would love for us to have a way to do runtime branding.  That said, the approach laid out in this OEP is probably just fine for the vast majority of Open edX operators except those who run large numbers of multi-tenant sites.  In that spirit, it feels like the right place to start, in my opinion.  Those with more complex needs are likely to need some sort of automation anyway, and are also more likely to have the resources/know how to implement it.  

Regarding runtime branding, as Adam said, there's a fundamental problem here that if the brand stylesheet is expressed as SCSS, it needs to be compiled into CSS at build time.  I haven't looked around to see if any packages exist to do it, but it's possible that that could be done quickly in the browser at runtime, but I'm not sure of the performance impact there.  

@abutterworth, another option I just thought of would be for brand stylesheets to be vanilla CSS only.  That would let us do runtime branding, I think.  We'd load the brand CSS after the base CSS and it'd just act as an override.  Could a brand have a build process where it creates a CSS file that includes all the classes that use all the SCSS variables it defines? 🤯  I'm not saying we should modify this OEP for this, but I wonder if there's something reasonable there...

There's another issue related to build/deployment pipelines for non-application frontend code, which we have no prior art on at the moment.  i.e., where do you deploy supporting libraries that are downloaded at runtime, and how does the MFE know to go get them?  Frontend pluggability will hopefully have some answers for this, as we fully expect that plugins will not be built as part of the MFEs, but will be independently deployed and loaded by the MFEs via a configuration document.

**Comment by davidjoy** [comment]
Do we need to say more in this OEP around how @edx/brand will be used in legacy frontends like edx-platform?  Or about how there are likely non-bootstrap areas of edx-platform which are incompatible with this approach, which we intend to leave behind?  Since a significant part of our frontend is still locked into the monolith, I wonder if it's worth describing in a bit more detail how this is intended to interact with comprehensive theming of those areas, which may be the only effective way of branding them today.  Maybe the answer there is that there are effectively two systems that can share styles if you want to put in the effort.

**Comment by OmarIthawi** [comment]
Adam and David thanks for the the answers. Yes, I understand that our needs are a bit on the extreme but what you've (David) suggested is a great compromise that could be helpful.

 - Would this OEP allow me to make a `appsembler/multi-tenant-brand` package that basically loads a CSS file at runtime without having this CSS file to be available at build time? This should cover our needs at the most basic level.

 - If I did so, I don't think it would be very different from injecting such CSS file in the MFE code itself, but without having a fork.

We'll figure out how to implement it in a multi-tenant manner.

**Comment by xitij2000** [comment]
> Regarding runtime branding, as Adam said, there's a fundamental problem here that if the brand stylesheet is expressed as SCSS, it needs to be compiled into CSS at build time. I haven't looked around to see if any packages exist to do it, but it's possible that that could be done quickly in the browser at runtime, but I'm not sure of the performance impact there.

It is possible to do in the browser at run time, but the sass compiler for browsers is over 800kb of JavaScript compressed, and might be low-performing, especially on mobiles, and especially if it needs to load dozens of other scss files to produce the final stylesheet(s).

File: oeps/oep-0048-brand-customization.rst (hunk 5 above)
Comment (abutterworth) [reply oeps/oep-0048-brand-customization.rst]: Noted :)

**Comment by abutterworth** [comment]
> Adam and David thanks for the the answers. Yes, I understand that our needs are a bit on the extreme but what you've (David) suggested is a great compromise that could be helpful.
> 
> * Would this OEP allow me to make a `appsembler/multi-tenant-brand` package that basically loads a CSS file at runtime without having this CSS file to be available at build time? This should cover our needs at the most basic level.
> * If I did so, I don't think it would be very different from injecting such CSS file in the MFE code itself, but without having a fork.
> 
> If this would happen figure out how to make the CSS site aware and does multi-tenant branding.

@OmarIthawi This OEP would not enable what you describe. The expectation here is that MFEs or edx-platform will consume the SCSS theme files to generate a CSS file for deployment. This brand package as described should have no executable code for accessing assets since doing so would create a requirement that the consuming application is running node js or python.

You could however, add a CSS import in the SCSS overrides theme file (refer to Paragon "SCSS Foundation" theming usage docs: https://edx.github.io/paragon/) that references a css file at some public URL.  Your `_overrides.scss` file in the brand package might look like:

```
@import url('https://mytenant.com/theme-override.css');
```

This solution isn't great. It would take longer to load and load extraneous css, but it would work. 

My take on this is that the solution to the multi-tenant theming problem will require more thought around runtime configuration of MFEs and the adoption of CSS-in-JS or CSS variables.

Please reach out if you'd like to chat over video call on this.

**Comment by bradenmacdonald** [comment]
@abutterworth @davidjoy 

Would something like this be compatible with what you have in mind / have built so far?

* A lightweight "MFE build service" runs and exposes a REST API to authorized users. (It could/should even be a serverless service running on AWS lambda or similar.)
* An Open edX operator opens up the "Theming MFE" at lms.example.com/admin/theming/ and changes the "branding" and "configuration" for all MFEs, and presses "Preview With [Profile]"
    - The Theming MFE makes a REST API call to the MFE build service
    - The MFE build service generates a `@example/brand-generated` npm package using the branding/configuration data passed in via the REST API and caches it
    - The MFE build service builds the `frontend-app-profile` in a docker container, with the `@example/brand-generated` package mounted. It is important that this takes on the order of seconds, not minutes.
    - The MFE build service pushes the newly built `frontend-app-profile` assets to an S3 bucket at some special preview URL like lms.example.com/profile-preview/    Perhaps the URL even changes every time, like lms.example.com/preview-mfe-387466278348/ .
    - The preview can be shown in an IFrame in the Theming MFE or in a new browser tab.
* The operator user adjusts the branding/configuration and repeats the preview until happy.
* The operator presses "Apply Theme" in the Theming MFE
    - The Theming MFE makes a REST API call to the MFE build service
    - The MFE build service rebuilds every MFE in parallel in docker containers, with the already-cached `@example/brand-generated` package mounted.
    - The MFE build service pushes the compiled MFE assets to S3+CloudFront, and everyone soon sees the new theme.

**Comment by xitij2000** [comment]
@bradenmacdonald This might need some discovery on how it'll work one of the options I wanted to explore was using CSS variables in a theme. The idea being that those who need compatibility with IE11 can use raw values, while others can use CSS variables as the input values into the theme, This might not work as-is right now since bootstrap pre-processes things like colours and produces lighter/darker variants, but it seems like bootstrap 5 is going in the css variables direction, so perhaps this will eventually be possible?

In this scenario, a css-variables-based theme could then be overridden by a styleseet that can be loaded dynamically to override all variables. Since this doesn't need to be recompiled it could work in seconds rather than a minute. 

Another thing idea that I think could be explored, is to allow mounting multiple branding packages. The default branding package could be installed via an NPM override, and the others could be installed as theme files in the theme directory. Which package is loaded can then be picked at runtime. In this scenario, the main theme file wouldn't be loaded as:
``import 'theme.scss'``
but as:
```javascript
let theme = dynamicallyGetThemeName();
import(`themes/${theme}/index.js`); 
```
My understanding is this will work even with dynamic imports, although it definitely needs to be tested. I did not find time this sprint to do that. 

In this scenario, the theme builder service can simple build a new package and inject it into an existing container. The themes directory itself could simply be a shared volume where the theme builder can place packages. But that is irrelevant to the approach itself.

**Comment by bradenmacdonald** [comment]
@xitij2000 I think those are cool ideas. That said, I'm wary of the amount of time and re-architecting may be required. Worth investigating though. My intuition right now is that something like what I outlined above will provide the best ROI since it just involves creating two new apps using existing patterns and not changing anything existing (I hope).

**Comment by xitij2000** [comment]
Yup, I think in any case there is definitely need for a UI tool to change the theme, and some system to rebuild the theme (whether it a single theme multiple themes or just a sheet with css vars)

**Comment by antoviaque** [comment]
FYI, one of the proposals from OpenCraft to the workshop is about MFE theming: https://openedx.atlassian.net/browse/OEROADMAP-13

The specific proposals we have for MFE theming there are related to our discussions here:

> 1.    The entire learner-facing UI will be converted to React-based MicroFrontEnds.
> 2.    Every MFE will support “branding” and “configuration” as defined at Theming Microfrontends
> 3.   Where useful and feasible, MFEs will also support “customization” and “frontend plugins” as defined in that same document.
> 4.  Open edX devstacks will come with two themes, “dark” and “edX”; the dark theme will be the default for Open edX in order to clearly show where theming is working correctly or not (seeing light colors on devstack or dark colors on edx.org prod means it’s not working). 
> 5.  Operators who wish to heavily customize the UI will be encouraged to create their own MFEs, and will be able to count on REST API stability during upgrades.
>     image
> 6.   Allow operators to change the theme quickly
>      *  A standard, automated workflow allows rebuilding any MFE with an updated theme configuration and deploying it [to S3+CloudFront] in under one minute.
>      * To allow a trial and error workflow for live editing of themes by users (see WYSIWYG Theming editors below), the changes should be viewable (or at least previewable) within a few seconds - quickly enough to allow to have a proper iterative editing workflow of themes from the web/WYSIWYG interface (edit -> see result -> edit -> etc.). It doesn't need to be instantaneous, but the compilation/application shouldn't take longer than it would be comfortable to a developer editing CSS.
> 7. Move the providers’ theming WYSIWYG editors upstream
>      * Reuse existing code from hosting providers if useful - e.g. OpenCraft's Instance Manager Ocim (and/or other providers who want to participate? Eg Appsembler's Tahoe, eduNext's Management Console, etc.)
>      * Theming WYSIWYG from these projects moves to become an official Open edX service/MFE, where people contribute to theming functionality and WYSIWYG editing together. Instead of isolated duplicate efforts, theming improvement and *maintenance* efforts (which can be huge!) can be mutualized into a shared community effort.
>      * The theming editor allows configuration from "within edX", i.e. closer to the way Wordpress allows theming settings from its admin interface. It allows simple theming operations (changing the “branding” and “configuration”), without requiring DevOps intervention/knowledge, or local proprietary modifications to deployment systems

CC @bradenmacdonald @xitij2000 @lgp171188

**Comment by abutterworth** [comment]
@bradenmacdonald This proposal is compatible with what you describe. How these brand packages get used will surely change over time. My hope here is that we can define a consistent location, name, and structure. That said, I do see one thing in your proposal that seems to be the likely sticking point:

> The MFE build service builds the frontend-app-profile in a docker container, with the @example/brand-generated package mounted. It is important that this takes on the order of seconds, not minutes.

I think MFEs currently take a couple minutes to build. Getting them down to seconds may need some work.

@xitij2000 I think CSS variables are the best future path for theming. Paragon is based on Bootstrap 4 and with an upgrade to Bootstrap 5 we can adopt CSS variables with that library. I'll investigate more in the coming months but I expect:

- This will enable real-time theming (by the nature of the CSS variables tech)
- Remove the need to compile SCSS all together (since they need to share variables) and instead allow the atomic delivery of Paragon or application styles.

**Comment by abutterworth** [comment]
Hi all, I have updated the status of this OEP to provisional. If you approve please provide a thumb and I'll merge it as provisional. We'll revisit this proposal after building out and proving an implementation.

**Review by davidjoy — APPROVED** [review]

**Review by xitij2000 — APPROVED** [review]
